### PR TITLE
allow concatenating "hybrid" (sparse/dense) tensors along their dense dimensions

### PR DESCRIPTION
**Pull request by umanwizard**
Follow-up to #13577 

The idea is to take each values tensor, concatenate it with zeros before and after itself (along the dimension corresponding to the one we're catting the tensors along), to get a tensor corresponding to the values for that tensor in the result. Then we concatenate all of those together to get the final values tensor. (Hopefully, this will be more clear from the example in the comments).

The indices are more straightforward: since we aren't concatenating along a sparse dimension, they don't change at all, so all we need to do are concatenate the indices from the different tensors together.

### CODE REVIEW

**Comment by umanwizard** [comment]
hmm... I thought my tests were passing, looks like I probably screwed something up here. Closing until I figure out how to fix it.

**Comment by umanwizard** [comment]
should be fixed now

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -61,61 +61,111 @@ static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except /* sho
   return true;
 }
 
+static void check_cat_sparse_dims(Tensor const &t,
```
Comment (weiyangfb) [inline aten/src/ATen/native/TensorShape.cpp]: can we have one sentence to explain what this function tries to do?

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -61,61 +61,111 @@ static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except /* sho
...
+      zeros_sizes[values_dim] = total_size - cumulative_size;
+      auto z2 = native::zeros(zeros_sizes, t._values().options());
+      vals_pieces.push_back(native::cat({z1, t._values(), z2}, values_dim));
+      idxs_pieces.push_back(t._indices());
     }
-    cumulative_offset += t.size(wrapped);
-    col += this_piece_size;
+    auto sizes_copy = sizes.vec();
+    sizes_copy[wrapped] = total_size;
+    return native::sparse_coo_tensor(native::cat(idxs_pieces, 1), native::cat(vals_pieces), sizes_copy, tensors[0].options());
```
Comment (weiyangfb) [inline aten/src/ATen/native/TensorShape.cpp]: does this mean catting a dense dim results to a uncoalesced sparse tensor? Should we add a note to that?

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -61,61 +61,114 @@ static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except /* sho
...
+  int64_t wrapped,
+  int64_t sparse_dim,
+  int64_t dense_dim) {
+    AT_CHECK(t.is_sparse(),
+            "Concatenating dense tensor at position ", pos, " with sparse tensor(s) not supported.");
+    AT_CHECK(sizes_match_except(sizes, t.sizes(), wrapped),
+            "Concatenating tensor at position ", pos, " of sizes ", t.sizes(), " with tensor of sizes ", sizes,
+            " along dimension ", dim, " not supported.");
+    AT_CHECK(t.sparse_dim() == sparse_dim && t.dense_dim() == dense_dim,
+            "Tensor at position ", pos, " has dimension: sparse ", t.sparse_dim(), ", dense ", t.dense_dim(),
```
Comment (weiyangfb) [inline aten/src/ATen/native/TensorShape.cpp]: and here, "All tensors must have same sparse_dim and dense_dim, but ..."

File: aten/src/ATen/native/TensorShape.cpp [hunk 4]
```diff
@@ -61,61 +61,114 @@ static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except /* sho
...
-      idxs[wrapped].narrow(0, col, this_piece_size) += cumulative_offset;
+  if (wrapped < sparse_dim) {
+    IntList sizes = tensors[0].sizes();
+    for (size_t i = 0; i < tensors.size(); ++i) {
+      auto const &t = tensors[i];
+      check_cat_sparse_dims(t, i, sizes, dim, wrapped, sparse_dim, dense_dim);
+      indices.push_back(t._indices());
+      values.push_back(t._values());
+    }
+    Tensor idxs = native::cat(indices, 1);
```
Comment (weiyangfb) [inline aten/src/ATen/native/TensorShape.cpp]: why not using `at::cat()` ?

**Review by weiyangfb — APPROVED** [review]
LGTM! Please address some minor nits

**Pull request closed without merging**